File: module-04-waiting/js/table_rows_wait.js
const { chromium } = require('playwright');

(async () => {
  const browser = await chromium.launch();
  const page    = await browser.newPage();
  await page.goto("https://example.com/table-ajax");

  // 1. 自訂輪詢
  const target = 15;
  const start  = Date.now();
  while (true) {
    const count = await page.$$eval("#data tr", els => els.length);
    if (count >= target) break;
    if (Date.now() - start > 20000) {
      throw new Error("Timeout: table rows did not reach target");
    }
    await page.waitForTimeout(500);
  }

  // 2. 輸出
  const finalCount = await page.$$eval("#data tr", els => els.length);
  console.log(`Total rows loaded: ${finalCount}`);

  await browser.close();
})();

File: module-04-waiting/py/table_rows_wait.py
from selenium import webdriver
from selenium.webdriver.chrome.service import Service
from webdriver_manager.chrome import ChromeDriverManager
from selenium.webdriver.common.by import By
import time

driver = webdriver.Chrome(
    service=Service(ChromeDriverManager().install()),
    options=webdriver.ChromeOptions().add_argument("--headless")
)
driver.get("https://example.com/table-ajax")

# 1. 自訂輪詢直到 <tr> 數量 >= 15
target = 15
start = time.time()
while True:
    rows = driver.find_elements(By.CSS_SELECTOR, "#data tr")
    if len(rows) >= target:
        break
    time.sleep(0.5)  # 每 0.5 秒再檢查一次
    if time.time() - start > 20:  # 最多等 20 秒
        raise TimeoutError("Table rows did not reach target in time")

# 2. 輸出
print(f"Total rows loaded: {len(rows)}")
driver.quit()
